File: lib/lexer.ml
open Token

type lexer =
  { input : string
  ; position : int
  ; read_position : int
  ; ch : char
  }
[@@deriving show]

let zero_char = char_of_int 0

let read_char { input; read_position; _ } =
  if read_position >= String.length input
  then
    { input
    ; position = read_position
    ; read_position = read_position + 1
    ; ch = zero_char
    }
  else
    { input
    ; position = read_position
    ; read_position = read_position + 1
    ; ch = String.get input read_position
    }
;;

let peak_char { input; read_position; _ } =
  if read_position >= String.length input
  then '\000'
  else String.get input read_position
;;

let is_letter c =
  let ascii_value = Char.code c in
  (ascii_value >= Char.code 'A' && ascii_value <= Char.code 'Z')
  || (ascii_value >= Char.code 'a' && ascii_value <= Char.code 'z')
  || c = '_'
;;

let read_identifier lexer =
  let rec aux lexer acc =
    if is_letter lexer.ch
    then aux (read_char lexer) (lexer.ch :: acc)
    else lexer, acc |> List.rev |> Utils.string_of_char_list
  in
  aux lexer []
;;

let is_digit c =
  let ascii_value = Char.code c in
  ascii_value >= Char.code '0' && ascii_value <= Char.code '9'
;;

let read_number lexer =
  let rec aux lexer acc =
    if is_digit lexer.ch
    then aux (read_char lexer) (lexer.ch :: acc)
    else lexer, acc |> List.rev |> Utils.string_of_char_list
  in
  aux lexer []
;;

let lookup_ident = function
  | "fn" -> Function
  | "let" -> Let
  | "true" -> True
  | "false" -> False
  | "if" -> If
  | "else" -> Else
  | "return" -> Return
  | _ -> Identifier
;;

let rec skip_whitespace l =
  if l.ch = ' ' || l.ch = '\t' || l.ch = '\r' || l.ch = '\n'
  then skip_whitespace @@ read_char l
  else l
;;

let next_token lexer =
  let lexer = skip_whitespace lexer in
  let { ch; _ } = lexer in
  match ch with
  (*operators*)
  | '=' ->
    if peak_char lexer = '='
    then read_char @@ read_char lexer, { token_type = Eq; literal = "==" }
    else read_char lexer, { token_type = Assign; literal = String.make 1 ch }
  | '+' -> read_char lexer, { token_type = Plus; literal = String.make 1 ch }
  | '-' -> read_char lexer, { token_type = Minus; literal = String.make 1 ch }
  | '*' ->
    read_char lexer, { token_type = Asterics; literal = String.make 1 ch }
  | '/' -> read_char lexer, { token_type = Slash; literal = String.make 1 ch }
  | '<' -> read_char lexer, { token_type = Lt; literal = String.make 1 ch }
  | '>' -> read_char lexer, { token_type = Gt; literal = String.make 1 ch }
  | '!' ->
    if peak_char lexer = '='
    then read_char @@ read_char lexer, { token_type = Not_eq; literal = "!=" }
    else read_char lexer, { token_type = Bang; literal = String.make 1 ch }
  | ';' ->
    read_char lexer, { token_type = Semicolon; literal = String.make 1 ch }
  | '(' -> read_char lexer, { token_type = Lparan; literal = String.make 1 ch }
  | ')' -> read_char lexer, { token_type = Rparan; literal = String.make 1 ch }
  | ',' -> read_char lexer, { token_type = Comma; literal = String.make 1 ch }
  | '{' -> read_char lexer, { token_type = Lbrace; literal = String.make 1 ch }
  | '}' -> read_char lexer, { token_type = Rbrace; literal = String.make 1 ch }
  | '\000' -> lexer, { token_type = Eof; literal = String.make 1 zero_char }
  | x when is_letter x ->
    let lexer, literal = read_identifier lexer in
    lexer, { token_type = lookup_ident literal; literal }
  | x when is_digit x ->
    let lexer, number = read_number lexer in
    lexer, { token_type = Integer; literal = number }
  | _ ->
    read_char lexer, { token_type = Illegal; literal = String.make 1 zero_char }
;;

let init input =
  read_char { input; position = 0; read_position = 0; ch = zero_char }
;;

File: go_rewrite/lexer/lexer.go
package lexer

import (
	tok "meow/token"
	"strings"
	"unicode"
)

type Lexer struct {
	Input    string
	Position int
}

func (l *Lexer) copy() *Lexer {
	l1 := Lexer{l.Input, l.Position}
	return &l1
}

func (l *Lexer) readChar() *Lexer {
	if len(l.Input) > int(l.Position) {
		l.Position++
	}
	return l
}

func (l *Lexer) char() byte {
	if l.Position >= len(l.Input) {
		return '\000'
	}
	return l.Input[l.Position]
}

func (l *Lexer) peekChar() byte {
	if l.Position+1 >= len(l.Input) {
		return '\000'
	}
	return l.Input[l.Position+1]
}

func (l *Lexer) skipWhitespace() {
	for l.char() == ' ' || l.char() == '\n' || l.char() == '\r' || l.char() == '\t' {
		l.readChar()
	}
}

func (l *Lexer) readInt() tok.Token {
	var buf strings.Builder
	for l.char() != '\000' && unicode.IsDigit(rune(l.char())) {
		buf.WriteByte(l.char())
		l.readChar()
	}
	return tok.Token{Token: tok.INTEGER, Literal: buf.String()}
}

func isLetter(c byte) bool {
	if (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || c == '_' {
		return true
	}
	return false
}

func lookupIdent(literal string) tok.Token {
	switch literal {
	case "fn":
		return tok.Token{Token: tok.FUNCTION, Literal: literal}
	case "let":
		return tok.Token{Token: tok.LET, Literal: literal}
	case "true":
		return tok.Token{Token: tok.TRUE, Literal: literal}
	case "false":
		return tok.Token{Token: tok.FALSE, Literal: literal}
	case "if":
		return tok.Token{Token: tok.IF, Literal: literal}
	case "else":
		return tok.Token{Token: tok.ELSE, Literal: literal}
	case "return":
		return tok.Token{Token: tok.RETURN, Literal: literal}
	default:
		return tok.Token{Token: tok.IDENTIFIER, Literal: literal}
	}
}

func (l *Lexer) readSymbol() string {
	var buf strings.Builder
	for l.char() != '\000' && (isLetter(l.char()) || unicode.IsDigit(rune(l.char()))) {
		buf.WriteByte(l.char())
		l.readChar()
	}
	literal := buf.String()
	return literal
}

func (l *Lexer) Next() tok.Token {
	for l.char() != '\000' {
		l.skipWhitespace()
		switch {
		case l.char() == '=':
			l.readChar()
			if l.char() == '=' {
				l.readChar()
				return tok.Token{Token: tok.EQ, Literal: "=="}
			}
			return tok.Token{Token: tok.ASSIGN, Literal: "="}
		case l.char() == '+':
			l.readChar()
			return tok.Token{Token: tok.PLUS, Literal: "+"}
		case l.char() == '-':
			l.readChar()
			return tok.Token{Token: tok.MINUS, Literal: "-"}
		case l.char() == '*':
			l.readChar()
			return tok.Token{Token: tok.ASTERICS, Literal: "*"}
		case l.char() == '/':
			l.readChar()
			return tok.Token{Token: tok.SLASH, Literal: "/"}
		case l.char() == '<':
			l.readChar()
			return tok.Token{Token: tok.LT, Literal: "<"}
		case l.char() == '>':
			l.readChar()
			return tok.Token{Token: tok.GT, Literal: ">"}
		case l.char() == '!':
			l.readChar()
			return tok.Token{Token: tok.BANG, Literal: "!"}
		case l.char() == ';':
			l.readChar()
			return tok.Token{Token: tok.SEMICOLON, Literal: ";"}
		case l.char() == '(':
			l.readChar()
			return tok.Token{Token: tok.LPAREN, Literal: "("}
		case l.char() == ')':
			l.readChar()
			return tok.Token{Token: tok.RPAREN, Literal: ")"}
		case l.char() == ',':
			l.readChar()
			return tok.Token{Token: tok.COMMA, Literal: ","}
		case l.char() == '{':
			l.readChar()
			return tok.Token{Token: tok.LBRACE, Literal: "{"}
		case l.char() == '}':
			l.readChar()
			return tok.Token{Token: tok.RBRACE, Literal: "}"}
		case unicode.IsDigit(rune(l.char())):
			return l.readInt()
		case isLetter(l.char()):
			symbol := l.readSymbol()
			return lookupIdent(symbol)
		default:
			return tok.Token{Token: tok.ILLEGAL, Literal: ""}
		}
	}
	return tok.Token{Token: tok.EOF, Literal: ""}
}
